README.md
# R Package GSE

GSE is an R package developed by Andy Leung for robust
estimation of multivariate location and scatter in the
presence of cellwise and casewise contamination and
missing data.

# Installation

Install the latest stable release from CRAN:

```
install.packages("GSE")
```

Alternatively, install the latest development version from GitHub:
```
install.packages("devtools")
devtools::install_github("acyleung/GSE")
```

# References


* Agostinelli, C., Leung, A., Yohai, V. J., & Zamar, R. H. (2015).
Robust estimation of multivariate location and scatter in the presence
of cellwise and casewise contamination. TEST, 24 (3), 441-461.
DOI: [10.1007/s11749-015-0450-6](http://dx.doi.org/10.1007/s11749-015-0450-6)

* Danilov, M., Yohai, V. J., & Zamar, R. H. (2012). Robust estimation of multivariate
location and scatter in the presence of missing data. J Amer Statist Assoc, 107,
1178-1186. DOI: [10.1080/01621459.2012.699792](http://dx.doi.org/10.1080/01621459.2012.699792)


ChangeLog
2017-04-15  Andy Leung  <andy.leung@stat.ubc.ca>
    * 4.1.1
    * Remove rrcov dependency

2016-12-26  Andy Leung  <andy.leung@stat.ubc.ca>
    * 4.1
    * Added filter option to TSGS function and removed filter related parameters options
    * Added DDC and UBF-DDC for filter choices 

2016-09-04  Andy Leung  <andy.leung@stat.ubc.ca>
    * 4.0
    * Added Rocke loss function option to GSE
    * Added cluster sampling option to EMVE
    * Added bivariate filtering option to GY-filter
    * Updated TSGS with the new options implemented
    * Separated the imputed S-estimator from GSE as an individual function

2016-08-04  Andy Leung  <andy.leung@stat.ubc.ca>
    * 3.3.1
    * Fixed mem problem in pmd calculation

2016-02-19  Andy Leung  <andy.leung@stat.ubc.ca>
    * 3.3
    * Fixed some error message handling
    * Update plot function according to changes made in ggplot2

2015-12-29  Andy Leung  <andy.leung@stat.ubc.ca>
    * 3.2.3
    * Added small perturbation to the point mass contamination in the data generator

2015-06-26  Andy Leung  <andy.leung@stat.ubc.ca>
    * 3.2.2
    * Increased the number of subsamples in EMVE.
    * Fixed seed problem

2015-06-26  Andy Leung  <andy.leung@stat.ubc.ca>
    * 3.2
    * Implemented fast 2SGS as described in the rejoinder of the 2SGS paper
    * Implemented the partial implementation procedure as described in the rejoinder of the 2SGS paper
    * Removed the accessor -- getInitial (unecessary)
    * Added 'rrcov' package in the imports field

2015-04-08  Andy Leung  <andy.leung@stat.ubc.ca>
    * 3.1.1
    * Fixed set seed issue
    * Fixed the condition number restriction bug in the estimation of EMVE

2014-08-04  Andy Leung  <andy.leung@stat.ubc.ca>
    * 3.1
    * Convert the iteration code of EM to all C++ (or Rcpp)
    * Convert the subsampling and concentration part of emve to all C++ (or Rcpp)
    * Include the initial estimates mu0 and S0 as an attribute of the GSE class
    * Include the imputed data matrix, weights, and first deriviative of weights used in the location estimation as new attributes of the GSE class
    * Create a new accessor -- getInitial -- for the new attributes of the GSE class (see above bullet)
    * Include TSGS, a new function that computes 2SGS estimate for cell- and case-wise outliers
    * Include Geochemical data that is analyzed in the 2SGS article

2013-06-04  Andy Leung  <andy.leung@stat.ubc.ca>
    * 3.0
    * Convert to S4 object system
    * Convert the iteration code of GSE to all C++ (or Rcpp)

2013-02-13  Andy Leung  <andy.leung@stat.ubc.ca>
    * 2.4.2
    * Fix the bug in rcondCov, now calculated condition number based on eigenvalues

2012-08-04  Andy Leung  <andy.leung@stat.ubc.ca>
    * 2.3
    * Include function to plot and print the result for different estimators
    * Fix the rand seed for emve C++ code
    * Update all the manuals

2012-07-15  Andy Leung  <andy.leung@stat.ubc.ca>
    * Fix the bug of asymetric covariance matrix when iteration is large
    * Algorithm stop when negative eigenvalue appears

2012-06-12  Andy Leung  <andy.leung@stat.ubc.ca>
    * 2.2
    * Include a line for only allowing up to dimension 50
    * Include GMME function
    * Extract cov.EM and slrt functions (and their related) from robmiss packages
    * No dependency on robmiss package

2012-06-10  Andy Leung  <andy.leung@stat.ubc.ca>
    * 2.1
    * Include the function to compute Quadrant Covariance
    * Include the function to compute Composite S-estimator
    * Include the version of GSE with initial estimator as quadrant covariance


2012-05-12  Andy Leung  <andy.leung@stat.ubc.ca>
    * 2.0
    * fixed bugs in the iterative step of GSE code, i.e. computing GSE based on fixed initial estimator
    * fixed bugs in computation of scale in GSE
    * include GSE with fixed point approach

2012-04-30  Andy Leung  <andy.leung@stat.ubc.ca>
    * 1.1
    * GSE is rewritten in C++ at least for the iterative step but not the loop
    * Reference manual for GSE is included
    * Faster implementation of both EMVE and GSE by performing the computation
      on groups with same missing patterns
    * Code for simulating contaminated data set is included

2012-03-17  Andy Leung  <andy.leung@stat.ubc.ca>
    * 1.0
    * Depends: R (>= 2.14.1), robmiss, Rcpp (>= 0.9.8), RcppArmadillo (>= 0.2.34)
    * Code for GSE computation is included
    * Faster implementation of both EMVE scale and partial mahalanobis distance
    * Computation of ML-estimate of location and scatter for each candidate within
      the resampling procedure is restricted to only 5 iterations

2012-02-14  Andy Leung  <andy.leung@stat.ubc.ca>
    * 0.1
    * Reference manual for mve.missing.C is included
    * Computation of scale for EMVE and partial mahalanobis distance is coded in C++

2012-01-14  Andy Leung  <andy.leung@stat.ubc.ca>
    * 0.0
    * Depends: R (>= 2.14.1), robmiss, Rcpp (>= 0.9.8), RcppArmadillo (>= 0.2.34)
    * Only include the function to compute EMVE
    * Resampling procedure is coded in C++
    * No code for GSE computation is included


R/TSGS.R

TSGS <- function(x, alpha=c(0.95,0.85), bivarQt=0.99, bivarCellPr=0.1, 
            partial.impute=FALSE, tol=1e-4, maxiter=150, method=c("bisquare","rocke"), 
            init=c("emve","qc","huber","imputed","emve_c"), mu0, S0, ...){
            
    xcall <- match.call()
    
    method <- match.arg(method)
    init <- match.arg(init)
    
    ## check dat
    if(is.data.frame(x) | is.matrix(x))
        x <- data.matrix(x)
    else stop("Data matrix must be of class matrix or data.frame.")
    if(any(is.na(x))) warning("Data matrix contains missing values.")
    ## June 12, 2012
    ## Only allow up to p=50
    n <- nrow(x)
    p <- ncol(x)
    if( p >200 | p < 2 ) stop("Column dimension of 'x' must be in between 2 and 200.")

    ## 1) filter step 
    xf <- gy.filt(x, alpha=alpha, bivarQt=bivarQt, bivarCellPr=bivarCellPr)

    ## 2) partial imputation step as suggested in the rejoinder of Agostinelli et al (2015)
    xf_pi <- xf
    if( partial.impute ){
        ximp <- .impute.coord.med(xf_pi)
        aid <- which(rowSums(!is.na(xf_pi)) == p)
        uid <- which(rowSums(!is.na(xf_pi)) < p)
        n0 <- n/2 + (p+1)
        if( n0 > length(aid) ){
            fid <- sample( uid, n0 - length(aid))
            xf_pi[fid,] <- ximp[fid,] 
        }
    }
    
    ## 3) estimation step
    res <- GSE(xf_pi, tol=tol, maxiter=maxiter, method=method, init=init, mu0, S0, ...)
    
    res <- new("TSGS",
        call = xcall,
        S = res@S,
        mu = res@mu,
        xf = xf,
        sc = res@sc,
        mu0 = res@mu0,
        S0 = res@S0, 
        iter = res@iter,
        eps = res@eps,
        estimator = "2SGS", 
        x = x,
        ximp = res@ximp,
        weights = res@weights,
        weightsp = res@weightsp,
        pmd = res@pmd,
        pmd.adj = res@pmd.adj,
        p = res@p,
        pu = res@pu)
    res
}


R/emve.R
emve <- function(x, maxits=5, sampling=c("uniform","cluster"), n.resample, n.sub.size, seed)
{
    xcall <- match.call()

    ## Check x
    if(is.data.frame(x) | is.matrix(x))
        x <- data.matrix(x)
    else stop("Data matrix must be of class matrix or data.frame")
    if( ncol(x) < 2 ) stop("Column dimension of 'x' must be at least 2.")

    ## Check n.resample and n.sub.size
    p <- ncol(x)
    sampling <- match.arg(sampling)
    ## EMVE setting
    if( missing(n.resample) & sampling == "uniform" ) n.resample <- 500    
    if( missing(n.sub.size) & sampling == "uniform" ) n.sub.size <- floor( (p+1)/(1-mean(is.na(x))) )
    ## EMVE_C setting
    if( missing(n.resample) & sampling == "cluster" ) n.resample <- 50    
    if( missing(n.sub.size) & sampling == "cluster" ) n.sub.size <- 2*floor( (p+1)/(1-mean(is.na(x))) )
    if( n.resample < 1 ) stop("Number of resampling must be >= 1.")
    if( n.sub.size < p+1 ) stop("Subsample size must be at least p+1.")
    
    ## Check seed
    if( missing(seed) ) seed <- 1000

    ## drop all rows with missing values (!!) :
    x_nonmiss <- is.na(x)*-1 + 1
    pp <- rowSums(x_nonmiss)
    pp_col <- colSums(x_nonmiss)
    
    ## Cannot contain all obs with completely missing rows!!
    if( all(pp == 0) ) stop("All observations have missing values!")
    if( any(pp_col == 0) )stop("Data matrix cannot contain column(s) with completely missing data!")

    ## Remove any rows with completely missing entries
    ok <- which(pp > 0)
    x_orig <- x
    x <- x[ ok,]
    x_nonmiss <- x_nonmiss[ ok,]
    
    ## reorder the data based on missingness
    x_sort <- .sort.missing(x, x_nonmiss)

    ## Check dimension 
    n <- nrow(x_sort$x); p <- ncol(x_sort$x)
    if(n <= p + 1)
        stop(if (n <= p) "n <= p -- you can't be serious!" else "n == p+1  is too small sample size")
    if(n < 2 * p)
        ## p+1 < n < 2p
        warning("n < 2 * p, i.e., possibly too small sample size")

    ## get initial estimate for all the EM calculation in EMVE subsampling
    mu0 <- colMeans(x_sort$x, na.rm=T)
    S0 <- diag(apply(x_sort$x, 2, var, na.rm=T))
    x_sort <- c(x_sort, .CovEM.setparam(p, mu0, S0))

    res <- with(x_sort, .emve.init(x, x_nonmiss, pu, n, p, theta, G.ind-1, length(theta),x.miss.group.match,
        miss.group.unique, miss.group.counts, miss.group.obs.col, miss.group.mis.col, 
        miss.group.p, miss.group.n, n.resample, n.sub.size, maxits, sampling, seed))

    S.chol <- tryCatch( chol(res$S), error=function(e) NA)
    if( !is.matrix(S.chol) )  stop("Estimated covariance matrix is not positive definite.")

    pmd <- pmd.adj <- rep(NA, nrow(x_orig))
    pmd[ok] <- res$pmd[x_sort$id.ro]
    pmd.adj[ok] <- res$pmd.adj[x_sort$id.ro]
    pu <- rowSums( !is.na(x_orig))
    res <- new("emve",
        call = xcall,
        S = res$S,
        mu = res$mu,
        sc=res$mve.scale,
        estimator = "Extended Minimum Volume Ellipsoid", 
        x = x_orig,
        pmd = pmd,
        pmd.adj = pmd.adj,
        p = p, 
        pu = pu)
}


## Assume the input data matrix is sorted using .sort.missing
.emve.init <- function(x, x_nonmiss, pu, n, p, theta0, G, d, x.miss.group.match, miss.group.unique, miss.group.counts, 
    miss.group.obs.col, miss.group.mis.col, miss.group.p, miss.group.n, n.resample, n.sub.size, maxits, sampling, seed)
{
    if( missing(sampling) ) sampling <- "uniform"
    
    ## EMVE setting
    if( missing(n.resample) & sampling == "cluster" ) n.resample <- 50  
    if( missing(n.sub.size) & sampling == "uniform" ) n.sub.size <- floor( (p+1)/(1-mean(is.na(x))) )
    ## EMVE_C setting
    if( missing(n.resample) & sampling == "cluster" ) n.resample <- 500     
    if( missing(n.sub.size) & sampling == "uniform" ) n.sub.size <- 2*floor( (p+1)/(1-mean(is.na(x))) )
    
    if( missing(maxits) ) maxits <- 5

    ## reset the seed, if there is one, upon exit
    if( missing(seed) ) seed <- 1000
    if (exists(".Random.seed", envir = .GlobalEnv, inherits = FALSE)) {
        seed.keep <- get(".Random.seed", envir = .GlobalEnv, inherits = FALSE)
        on.exit(assign(".Random.seed", seed.keep, envir = .GlobalEnv))
    }
    set.seed(seed)
    
    ## Initial scales for later calculation
    scale0_init <- .scale.mve.init(pu)  # will be used to calculate EMVE scale
    cc <- scale0_init$cc
    ck <- scale0_init$ck
    colmed <- apply(x, 2, median, na.rm=T) # coordinate-wise medians for each column, full data

    ## Filled each entry by the overall median for that variable
    x_filled <- x
    a <- is.na(x)%*%diag(colmed)
    x_filled[is.na(x_filled)] <- 0
    x_filled <- x_filled + a
    
    res <- .emve.Rcpp(x_filled, x_nonmiss, pu, n, p, theta0, G, d, x.miss.group.match, miss.group.unique, miss.group.counts,
            miss.group.obs.col, miss.group.mis.col, miss.group.p, miss.group.n, n.resample, n.sub.size, sampling, cc, ck, maxits)
    res <- list(
        S = res$S0,
        mu = res$mu0,
        mve.scale=res$ss0)
        
    x.tmp <- sweep(x, 2, res$mu, "-")
    pmd <- .partial.mahalanobis.Rcpp(x.tmp, res$S, miss.group.unique, miss.group.counts)
    
    ## compute adjusted pmd
    pmd.adj <- qchisq( pchisq( pmd, df=pu, log.p=T, lower.tail=F), df=p, log.p=T, lower.tail=F) 
    pmd.adj[ which( pu == p)] <- pmd[ which(pu==p) ]

    ## size adjust
    cf <- median( pmd.adj )/qchisq(0.5, df=p)
    res$S <- cf * res$S
    pmd <- pmd / cf
    pmd.adj <- pmd.adj / cf
    
    res$pmd <- pmd
    res$pmd.adj <- pmd.adj
    res
}


R/filter.R
.gy.filt.uni <- function(v, alpha, df){
    n <- length(v)
    id <- (1:n)[!is.na(v)]
    v.out <- rep(NA, n)
    
    v <- na.omit(v)
    n <- length(v)
    v.order <- order(v)	
    v <- sort(v)
    i0 <- which(v < qchisq( alpha, df ))
    n0 <- 0
    if( length(i0) > 0){
        i0 <- rev(i0)[1]
        dn <- max( pmax( pchisq( v[i0:n], df) - (i0:n - 1)/n, 0)) 
        n0 <- round(dn*n)
    } 
    v <- v[ order(v.order) ] 
    v.na <- v
    if(n0 > 0) v.na[ v.order[ (n - n0 + 1):n] ] <- NA
    
    v.out[id] <- v.na
    
    return( v.out )
}

.gy.filt.uni.it <- function(v, alpha, df, miter){
    converge <- 0
    iter <- 0
    n <- length(v)
    n_flag <- sum(is.na(v))
    while( converge == 0 & iter < miter ){
        iter <- iter + 1
        v <- .gy.filt.uni( v, alpha, df )
        n_flag_new <- sum(is.na(v))
        if( n_flag_new - n_flag <= 0 ) converge <- 1
    }
    return( v)
}

gy.filt <- function(x, alpha=c(0.95,0.95), bivarQt=0.99, bivarCellPr=0.1, miter=5){

    if( missing(alpha) ) alpha <- c(0.95, 0)

    x <- as.matrix(x)
    n <- nrow(x)
    p <- ncol(x)
    xs <- scale(x, apply(x, 2, median, na.rm=TRUE), apply(x, 2, mad, na.rm=TRUE))
    u <- 1-is.na(x)
    if( alpha[1] > 0) u <- 1-is.na(apply(xs^2, 2, .gy.filt.uni.it, alpha=alpha[1], df=1, miter=miter))
    m <- matrix(0, n, p)
    mn <- matrix(0, n, p)
    if( alpha[2] > 0){
        for(j in 1:(p-1))
            for(k in (j+1):p){
                S <- matrix(1, 2, 2)
                S[1,2] <- S[2,1] <- (mad( na.omit(xs[,j] + xs[,k]) )^2 - 
                                            mad( na.omit(xs[,j] - xs[,k]) )^2)/4

                pmd <- mahalanobis(xs[,c(j,k)], rep(0,2), S)
                pmd[ which(u[,j]*u[,k] == 0) ] <- -1
                pmd <- pmd * qchisq(0.5, df=2) / median(pmd[which(pmd > 0)])
                mn[,c(j,k)] <- mn[,c(j,k)] + 1*(pmd > 0)
                pmd[which(pmd > 0)] <- .gy.filt.uni.it(pmd[which(pmd > 0)], alpha=alpha[2], df=2, miter=miter)
                m[,c(j,k)] <- m[,c(j,k)] + 1*is.na(pmd)
            }
        u[m > qbinom(bivarQt, mn, prob=bivarCellPr) & mn > 0] <- 0
    }
    xna <- x
    xna[u == 0] <- NA
    return( xna )
}

R/emve-utils.R
.emve.Rcpp <- function(x, x_nonmiss, pu, n, p, theta0, G, d, x.miss.group.match, miss.group.unique, miss.group.counts,
        miss.group.obs.col, miss.group.mis.col, miss.group.p, miss.group.n, nresample, nsubsize, sampling, cc, ck, EM_maxits){

    x_na <- x
    x_na[x_nonmiss == 0] <- NA
    subsample.id <- 0:(n-1)
    if(sampling == "cluster") subsample.id <- .sample.cluster(x_na) - 1
    
    res <- tryCatch( .Call("emve_Rcpp", x, x_nonmiss, pu, n, p, theta0, G, d, x.miss.group.match, miss.group.unique, miss.group.counts,
        miss.group.obs.col, miss.group.mis.col, miss.group.p, miss.group.n, nresample, nsubsize, subsample.id, cc, ck, EM_maxits),
        "std::range_error" = function(e){
        conditionMessage( e ) } )

    if( is.character( res ) )
        stop( paste(res, "\nEstimated covariance matrix is not positive definite.\n") )

    ss0 <- res[1,1,]
    mu0 <- t(res[2,,])
    S0 <- res[-c(1,2),,]

    list(ss0=ss0, mu0=mu0, S0=S0)
}

###################################################################
## Compute the correction constants k and 50% chi-sq quantiles for 
## the computation of the scale of MVE for incomplete data (new version)
.scale.mve.init <- function(pp){
    cc <- qchisq(1/2,pp)
    k1 <- 1/pp
    k2 <- 0.5^(pp/2)
    k3 <- 1/(gamma(pp/2))
    k4 <- cc^(1+pp/2)
    k5 <- exp(-cc/2)
    kk <- k1 * k2 * k3 * k4 * k5
    ck <- cc*kk
    return( list(cc=cc, ck=ck) )
}


###################################################################
## Partial mahalanobis distance
## Rcpp version
.partial.mahalanobis.Rcpp <- function(x_mu_diff, Sigma, miss_group_unique, miss_group_counts){
    res <- tryCatch( .Call("fast_partial_mahalanobis", x_mu_diff, Sigma, miss_group_unique, miss_group_counts),
        "std::range_error" = function(e){
        conditionMessage( e ) } )
    if( is.character( res ) ) stop(res)
    return( c(res) )
}


###################################################################
## clustering-based sampling
.sample.cluster <- function(x, explvar = 0.9999, method.clust="ward.D", scalefn=Qn)
{
    n <- nrow(x)
    p <- ncol(x)
    x.loc <- apply(x, 2, median, na.rm=TRUE)
    x.scale <- apply(x, 2, function(x) scalefn(na.omit(x)))
    if(any(x.scale == 0)) stop("More than 50% equal values in one or more variables!")
    x.sc <- scale(x, x.loc, x.scale)
    R <- matrix(1, p, p)
    for (i in 1:(p-1)) for (j in (i+1):p) {
        R[j, i] <- R[i, j] <- (scalefn( na.omit(x.sc[, i] + x.sc[, j]) )^2 - 
                                 scalefn( na.omit(x.sc[, i] - x.sc[, j]) )^2)/4
    }
    if(any(is.na(x))) x.sc <- .impute.simple(x, rep(0,p))
    R.eigen <- eigen(R)
    R.eigen.L <- R.eigen$values
    p0 <- rev(which(R.eigen.L > 0))[1]
    R.eigen.L[which(R.eigen.L < 0)] <- 0
    p1 <- (1:p0)[(cumsum(R.eigen.L)/sum(R.eigen.L) > explvar)][1]
    x.pc <- x.sc %*% R.eigen$vectors[, 1:p1]
    xpc.sc <- scale(x.pc, apply(x.pc, 2, median), apply(x.pc, 2, scalefn))
    hctree <- hclust(dist(xpc.sc), method=method.clust)
    cont <- TRUE
    b <- 1
    while(cont){
        b <- b + 1
        clusterB <- cutree(hctree, k=b)
        cont <- !all(table(clusterB) < n/2)
    }
    b <- b-1
    clusterB <- cutree(hctree, k=b)
    clusterB_sizes <- table(clusterB)
    clusterB1 <- which( clusterB_sizes == max(clusterB_sizes))
    clusterB1.id <- which(clusterB %in% clusterB1)

    return(clusterB1.id)  
}


R/GSE.R
###########################################################
## Generalized S-Estimator
###########################################################
GSE <- function(x, tol=1e-4, maxiter=150, method=c("bisquare","rocke"), init=c("emve","qc","huber","imputed","emve_c"), mu0, S0, ...)
{
    xcall <- match.call()

    ## argument checks
    method <- match.arg(method)
    init <- match.arg(init)

    ## check dat
    if(is.data.frame(x) | is.matrix(x))
        x <- data.matrix(x)
    else stop("Data matrix must be of class matrix or data.frame")
    
    p <- ncol(x)
    if( p >200 | p < 2 ) stop("Column dimension of 'x' must be in between 2 and 200.")
    
    ## drop all rows with missing values (!!) :
    x_nonmiss <- is.na(x)*-1 + 1
    pp <- rowSums(x_nonmiss)
    pp_col <- colSums(x_nonmiss)
    
    ## Cannot contain all obs with completely missing rows!!
    if( all(pp == 0) ) stop("All observations have missing values!")
    if( any(pp_col == 0) )stop("Data matrix cannot contain column(s) with completely missing data!")	
    
    ok <- which(pp > 0)
    x_orig <- x
    x <- x[ ok,]
    x_nonmiss <- x_nonmiss[ ok,]

    ## reorder the data based on missingness
    x_sort <- .sort.missing(x, x_nonmiss)

    ## get initial estimate for all the EM calculation in EMVE subsampling
    EM.mu0 <- colMeans(x_sort$x, na.rm=T)
    EM.S0 <- diag(apply(x_sort$x, 2, var, na.rm=T))
    x_sort <- c(x_sort, .CovEM.setparam(p, EM.mu0, EM.S0))
    
    ## dimension
    n <- nrow(x_sort$x); p <- ncol(x_sort$x)
    if(n <= p + 1)
        stop(if (n <= p) "n <= p -- you can't be serious!" else "n == p+1  is too small sample size")
    if(n < 2 * p)
        ## p+1 < n < 2p
        warning("n < 2 * p, i.e., possibly too small sample size")

    if( xor(missing(mu0), missing(S0)) ) warning("Both 'mu0' and 'S0' must be provided. Default 'init' is used...")
    if( missing(mu0) || missing(S0) ){
        init.res <- switch( init,
            emve= {res <- emve(x, sampling="uniform", ...); list(mu=res@mu, S=res@S) },
            emve_c= {res <- emve(x, sampling="cluster", ...); list(mu=res@mu, S=res@S) },
            qc ={res <- HuberPairwise(x, psi="sign", computePmd = FALSE); list(mu=res@mu, S=res@S)},
            huber = {res <- HuberPairwise(x, psi="huber", computePmd = FALSE, ...); list(mu=res@mu, S=res@S)},
            imputed = {ximp_simp <- .impute.simple(x, apply(x, 2, median, na.rm=TRUE)); 
                    res <- rrcov::CovSest(ximp_simp, method=method);
                    list(mu=res@center, S=res@cov) }
            )
        S0 <- init.res$S
        mu0 <- init.res$mu
    } 
    S0.chol <- tryCatch( chol(S0), error=function(e) NA)
    if( !is.matrix(S0.chol) )  stop("Estimated initial covariance matrix 'S0' is not positive definite.")

    ## initiate GSE computation
    bdp <- 0.5
    print.step <- 0
    tol.scale <- 1e-9
    miter.scale <- 300
    res <- with(x_sort, .GSE.init(x, x_nonmiss, bdp, pu, n, p, miss.group.unique, miss.group.counts, mu0, S0, tol, 
        maxiter, tol.scale, miter.scale, print.step=print.step, method))
    
    ## compute pmd
    pmd <- pmd.adj <- rep(NA, nrow(x_orig))
    pmd[ok] <- res$pmd[x_sort$id.ro]
    pmd.adj[ok] <- res$pmd.adj[x_sort$id.ro]
    
    ## new 2014-07-28
    wgts <- rep(NA, nrow(x_orig))
    wgts[ok] <- res$weights[x_sort$id.ro]
    wgtsp <- rep(NA, nrow(x_orig))
    wgtsp[ok] <- res$weightsprm[x_sort$id.ro]
    ximp <- matrix(NA, nrow(x_orig), ncol(x_orig))
    ximp[ok,] <- res$ximp[x_sort$id.ro,]
    
    res <- new("GSE",
        call = xcall,
        S = res$S,
        mu = res$mu,
        sc = res$stilde0,
        mu0 = mu0,
        S0 = S0, 
        iter = res$iter,
        eps = res$ep,
        estimator = "Generalized S-Estimator", 
        x = x_orig,
        ximp = ximp,
        weights = wgts,
        weightsp = wgtsp,
        pmd = pmd,
        pmd.adj = pmd.adj,
        p = p,
        pu = pp)
    res
}

## Assume the input data matrix is sorted using .sort.missing
.GSE.init <- function(x, x_nonmiss, bdp, pu, n, p, miss.group.unique, miss.group.counts,  mu0, S0, tol, maxiter, tol.scale, miter.scale, print.step, method)
{
    ##########################################################################################
    ## computing
    p.const.group <- rowSums(miss.group.unique)
    if( method == "bisquare"){
        tuning.const.group <- .rho.bisquare.tune(p.const.group, bdp)
        res <- .GSE.Rcpp(x, matrix(mu0,1,p), S0, tol, maxiter, tol.scale, miter.scale, 
            miss.group.unique, miss.group.counts, tuning.const.group, print.step, bdp)
    }
    if( method == "rocke"){
        tuning.const.group <- .rho.rocke.tune(p.const.group)
        gamma.const.group <- pmin(1, qchisq(0.95, df=p.const.group)/p.const.group - 1)
        res <- .GSE.rocke.Rcpp(x, matrix(mu0,1,p), S0, tol, maxiter, tol.scale, miter.scale, 
            miss.group.unique, miss.group.counts, tuning.const.group, gamma.const.group, print.step, bdp)
    }
    ##########################################################################################
    ## Include additional output other than mu and S output by .GSE.fixpt.init:
    ## input data 'x', nonmissing variables 'pu', partial MD 'pmd', 
    ## note: need to reorder x, pmd, pu
    
    ## Compute pmd
    x.tmp <- sweep(x, 2, c(res$mu), "-")
    pmd <- .partial.mahalanobis.Rcpp(x.tmp, res$S, miss.group.unique, miss.group.counts)
    pmd.adj <- qchisq( pchisq( pmd, df=pu, log.p=T, lower.tail=F), df=p, log.p=T, lower.tail=F) 
    pmd.adj[ which( pu == p)] <- pmd[ which(pu==p) ]
    res$pmd <- pmd
    res$pmd.adj <- pmd.adj

    ## new 2016-08
    pmd.adj.med <- median(pmd.adj)
    qchis5 <- qchisq(.5,p)
    res$S <- (pmd.adj.med/qchis5)*res$S
    pmd <- (qchis5/pmd.adj.med)*pmd
    pmd.adj <- (qchis5/pmd.adj.med)*pmd.adj
    
    ##########################################################################################
    ## Output
    res
}





